internal/definitions/definition_handler_identity_claim.go

```diff
@@ -0,0 +1,218 @@
+			if !dh.getSystemBroadcastPayload(ctx, msg, data, &verification) {
+				return nil, nil
+			}
+			identityMatches = verification.Identity.Equals(ctx, &identity.IdentityBase)
+			verificationID = verification.Claim.ID
+			verificationHash = verification.Claim.Hash
+			if identityMatches && msg.Header.ID.Equals(verificationID) && msg.Hash.Equals(verificationHash) {
+				return candidate.Header.ID, nil
+			}
+		}
+		log.L(ctx).Warnf("Skipping invalid potential verification '%s' for identity claimID='%s' claimHash=%s: foundData=%t identityMatch=%t id=%s hash=%s", candidate.Header.ID, msg.Header.ID, msg.Hash, foundAll, identityMatches, verificationID, verificationHash)
+	}
+	return nil, nil
+}
+
+func (dh *definitionHandlers) handleIdentityClaim(ctx context.Context, state DefinitionBatchState, msg *fftypes.Message, identityClaim *fftypes.IdentityClaim, verificationID *fftypes.UUID) (DefinitionMessageAction, error) {
+	l := log.L(ctx)
+
+	identity := identityClaim.Identity
+	parent, retryable, err := dh.identity.VerifyIdentityChain(ctx, identity)
+	if err != nil && retryable {
+		return ActionRetry, err
+	} else if err != nil {
+		// This cannot be processed as the parent does not exist (or similar).
+		// We treat this as a bad request, as nodes should not be broadcast until the parent identity is
+		// is already confirmed. (Note different processing for org/custom childs, where there's a parent
+		// verification to coordinate).
+		l.Warnf("Unable to process identity claim %s: %s", msg.Header.ID, err)
+		return ActionReject, nil
+	}
+
+	// Check signature verification
+	if !dh.verifyClaimSignature(ctx, msg, identity, parent) {
+		return ActionReject, nil
+	}
+
+	existingIdentity, err := dh.database.GetIdentityByName(ctx, identity.Type, identity.Namespace, identity.Name)
+	if err == nil && existingIdentity == nil {
+		existingIdentity, err = dh.database.GetIdentityByID(ctx, identity.ID)
+	}
+	if err != nil {
+		return ActionRetry, err // retry database errors
+	}
+	if existingIdentity != nil && !existingIdentity.IdentityBase.Equals(ctx, &identity.IdentityBase) {
+		// If the existing one matches - this is just idempotent replay. No action needed, just confirm
+		l.Warnf("Unable to process identity claim %s - conflict with existing: %v", msg.Header.ID, existingIdentity.ID)
+		return ActionReject, nil
+	}
+
+	// Check uniquness of verifier
+	verifier := dh.getClaimVerifier(msg, identity)
+	existingVerifier, err := dh.database.GetVerifierByValue(ctx, verifier.Type, identity.Namespace, verifier.Value)
+	if err != nil {
+		return ActionRetry, err // retry database errors
+	}
+	if existingVerifier != nil && !existingVerifier.Identity.Equals(identity.ID) {
+		log.L(ctx).Warnf("Unable to process identity claim %s - verifier type=%s value=%s already registered: %v", msg.Header.ID, verifier.Type, verifier.Value, existingVerifier.ID)
+		return ActionReject, nil
+	}
+
+	if parent != nil && identity.Type != fftypes.IdentityTypeNode {
```

Comment:
Struggling to follow this block - it only executes for non-root, non-node identities? Could be worth an extra comment...

Reply:
After reading more of the code, it does confirm my understanding that root identities and nodes have self-signed claims and do not require a separately signed verification message. But they still require "verifiers"? I think I'm very close to understanding, but not quite there.

Reply:
> root identities 

They have self-signed claims. The `messages.verifier` entry will be empty.

> nodes have self-signed claims 

They have claims signed by the `parent` directly at this time, as Data Exchange keys have not been made first class such that they can sign a claim. Only blockchain keys can do that at this time. The `messages.verifier` entry will be empty.